File: iosApp/Sources/Analytics/AnalyticsRows.swift
import Foundation
import Shared

/// Which of the drawer's own sections a row belongs to, in display order.
/// Port of Android's `AnalyticsGroup`.
enum AnalyticsGroup: CaseIterable {
    case localVideos
    case ownedMatches
    case shared

    var label: String {
        switch self {
        case .localVideos:  return "On this phone"
        case .ownedMatches: return "My matches"
        case .shared:       return "Shared with me"
        }
    }
}

/// What an `analysable` row's control should show, in place of a live "Analyze"
/// button, when the pipeline is already busy with its video or its last attempt
/// failed. Port of Android's `AnalyseAffordance`.
enum AnalyseAffordance: Equatable {
    /// Nothing is running: the button is live.
    case ready
    /// A run is under way. `phase` is the drawer's own wording for that stage.
    case inProgress(phase: String)
    /// The last attempt ended in a failure, with the pipeline's own message.
    case failed(reason: String)
    /// A device run stopped because the app left the foreground.
    ///
    /// A fourth case Android does not have, and the one place this list
    /// deliberately diverges: over there a foreground service carries a run
    /// through a locked screen, and here nothing can. Folding it into `.failed`
    /// would put "Retry" on a run that did not go wrong and colour the line red;
    /// folding it into `.inProgress` would spin a control over something that
    /// has stopped. See `LocalAnalysisState.paused`.
    case paused(reason: String)
}

/// Where an `analysable` row's button goes when it is tapped.
enum AnalyseAction: Equatable {
    /// A failed cloud run whose court is already saved: `AnalyzeCoordinator.retry`
    /// picks up at the step that failed, so there is nothing to ask the coach for.
    case resume(entryId: String)
    /// Everything else starts at court marking.
    case markCourt(entryId: String)
}

/// One row of the Analytics list. `entryId` is nil exactly when there is no video
/// on this phone behind the row, which is also the only case that is inert.
struct AnalyticsRow: Identifiable, Equatable {
    let key: String
    let entryId: String?
    let group: AnalyticsGroup
    let title: String
    let subtitle: String
    let state: AnalyticsRowState
    let affordance: AnalyseAffordance

    var id: String { key }
}

/// The one explanatory line above the list, chosen from the row states actually
/// on screen rather than from the platform.
///
/// Four cases. Android had two - all-inert, or the dot legend - which put a dot
/// legend over a list with no dots the moment anything was analysable. That was
/// found here, where every local video was analysable while no track store
/// existed, and ported back there once it turned out to be a live defect on
/// Android too.
///
/// Both of the middle cases stay ordinary now that a track store does exist: a
/// coach whose matches were all filmed elsewhere sees an inert list, and a coach
/// with videos on the phone and nothing analysed yet sees live "Analyze" buttons
/// and no dot to explain. Explaining a dot that cannot render is the same defect
/// as a comment outliving its code.
enum AnalyticsLegend: Equatable {
    /// The list is empty; there is nothing to explain. Not spelled `none`: the
    /// moment anything holds an `AnalyticsLegend?`, `.none` binds to
    /// `Optional.none` instead and the mistake is silent.
    case silent
    /// Every row is inert, all for the same reason.
    case nothingOnThisPhone
    /// At least one row carries the availability dot.
    case dot
    /// No dots, but at least one live "Analyze" button.
    case analyseButton

    /// Whether any row draws the availability dot the `.dot` line explains.
    var showsDot: Bool { self == .dot }

    /// Whether an inert row spells out "Not on this phone" for itself. False only
    /// when the legend already said it once for the whole list, so the legend and
    /// the rows cannot disagree about which of them is doing the explaining.
    var showsNotOnDeviceSubtitle: Bool { self != .nothingOnThisPhone }

    var text: String? {
        switch self {
        case .silent:
            return nil
        case .nothingOnThisPhone:
            return "None of these matches are on this phone yet."
        case .dot:
            // "- tap to view", as Android has it. The words were dropped while
            // there was no detail screen to open and nothing to tap; both exist
            // now, so promising it is accurate again.
            return "Analyzed on this phone - tap to view"
        case .analyseButton:
            // What the button actually does, said plainly, so a screen called
            // Analytics does not look like it is about to draw a chart. Both
            // targets named on purpose: court marking offers cloud AND on
            // device, and it is the on-device run that writes the track a row
            // needs to turn ready, so copy naming only the cloud would steer a
            // coach away from the dot this same legend explains.
            return "Analyze cuts a video into rallies, on this phone or in the cloud."
        }
    }
}

/// What either pipeline is doing with this video right now.
///
/// TWO liveness sources, composed, as Android's `affordanceFor` composes them.
/// A device run never moves `LocalVideoEntry.stage`, so the stage-based rules in
/// shared cannot see one at all: a row reading the cloud stage alone would show
/// a live "Analyze" button over a run already in progress, and pressing it walks
/// the coach through marking the court again and then returns immediately
/// because the entry is already running. Android shipped exactly that - twelve
/// taps, no effect, no explanation - which is why `isDeviceRunInFlight` exists
/// and why it is consulted here.
///
/// The device speaks first when it has anything to say. It is the pipeline this
/// phone controls, and its states are the ones that end in a track the row can
/// then offer; a cloud stage left over from an earlier attempt must not describe
/// a device run happening now.
///
/// The phase wording is shared on both sides: `deviceWorkLabel` for the device
/// run, `LocalVideoStatus` for the cloud one, so this row and the chrome
/// indicator over the same video say the same words.
func analyseAffordance(
    for entry: LocalVideoEntry,
    progress: AnalyzeProgress?,
    device: LocalAnalysisState
) -> AnalyseAffordance {
    switch device {
    case .preparing, .analysing, .cutting:
        // Non-nil for every running state, since toDeviceWork returns nil only
        // for the ones this branch excludes.
        if let work = toDeviceWork(entryId: entry.id, state: device) {
            return .inProgress(phase: BackgroundWorkKt.deviceWorkLabel(
                phase: work.phase, fraction: work.fraction
            ))
        }
        return .inProgress(phase: "Analyzing on device")
    case .failed(let message):
        return .failed(reason: message)
    case .paused:
        // The percentage only when there is one: a run frozen while preparing,
        // or while cutting clips, has no analysis fraction to quote and must not
        // invent a zero.
        guard let fraction = device.pausedFraction else {
            return .paused(reason: "Paused - keep Shuttl open to finish")
        }
        return .paused(reason: "Paused at \(Int(fraction * 100))% - keep Shuttl open to finish")
    case .idle, .done:
        break
    }
    if LocalVideoStatus.isRunning(stage: entry.stage) {
        // Non-nil for both running stages, so the fallback is a guard against a
        // future stage rather than a case anything reaches today.
        let phase = LocalVideoStatus.text(stage: entry.stage, progress: progress)
        return .inProgress(phase: phase ?? "Analyzing…")
    }
    if entry.stage == .failed {
        return .failed(reason: entry.failureMessage ?? "Unknown error")
    }
    return .ready
}

/// Where this row's button goes. The rule itself is shared
/// (`canResumeFailedAnalysis`) rather than written out here: five hand-written
/// copies of `stage == .failed && keypoints != nil` are exactly what put it in
/// `shared/` in the first place, and this would have been the sixth.
func analyseAction(for entry: LocalVideoEntry) -> AnalyseAction {
    LocalVideoEntryKt.canResumeFailedAnalysis(entry: entry)
        ? .resume(entryId: entry.id)
        : .markCourt(entryId: entry.id)
}

/// How a local video describes itself under its title: how long it runs and when
/// it was added.
///
/// Shared with the drawer's own local video row rather than written out twice.
/// The two screens list the same videos side by side in one tap, and a comment
/// claiming they cannot describe a video two ways is worth less than a call that
/// makes it so.
///
/// Sentence case, as the mock has it ("1:05 · Sep 2"). Neither this screen nor
/// the drawer uppercases it any more: the redesign carries no uppercase
/// anywhere, and Analytics kept transforming it at its own call site for a
/// while after the drawer had stopped.
func localVideoSubtitle(_ entry: LocalVideoEntry) -> String {
    let duration = LocalVideoLogic.formatDuration(ms: entry.durationMs)
    let date = formatMatchDate(millis: entry.addedAtEpochMs)
    return "\(duration) · \(date)"
}

/// Builds the Analytics list's rows, grouped like the drawer: local videos, then
/// owned matches, then shared. `analyticsRowState` alone decides READY /
/// ANALYSABLE / NOT_ON_DEVICE; this only gathers its two booleans per match and,
/// for an analysable one, reads the cloud pipeline's liveness through
/// `analyseAffordance` so a row can show progress or a failure instead of a live
/// button. `hasStoredTrack` alone cannot tell a run in flight, or one that just
/// failed, from one never attempted.
///
/// Port of Android's `buildAnalyticsRows`, now including its `storedTrackIds`
/// and its on-device liveness map.
func buildAnalyticsRows(
    localEntries: [LocalVideoEntry],
    ownedRows: [MatchRow],
    sharedMatches: [MatchSummary],
    progressByEntryId: [String: AnalyzeProgress],
    /// Entries with a track this build can draw. Passed in as a set rather than
    /// probed per row: the store answers from a file header, and a list of forty
    /// matches would ask forty times on every redraw.
    storedTrackIds: Set<String> = [],
    /// What the on-device runner is doing, per entry. Absent means idle.
    deviceStates: [String: LocalAnalysisState] = [:]
) -> [AnalyticsRow] {
    func rowFor(
        key: String, entryId: String?, group: AnalyticsGroup, title: String, subtitle: String
    ) -> AnalyticsRow {
        let entry = entryId.flatMap { id in localEntries.first { $0.id == id } }
        let state = analyticsRowState(
            hasLocalEntry: entry != nil,
            hasStoredTrack: entry.map { storedTrackIds.contains($0.id) } ?? false
        )
        var affordance = AnalyseAffordance.ready
        // Only an analysable row reads liveness. A ready row keeps its dot while
        // a second run is in flight, because the track the first run produced is
        // still there; and a not-on-device row has no run of its own to report.
        //
        // `entry` is never nil in this branch: analysable requires a local entry,
        // which is exactly `entry != nil` above. The binding is a guard, not a
        // second source of truth.
        if state == .analysable, let entry {
            affordance = analyseAffordance(
                for: entry,
                progress: progressByEntryId[entry.id],
                device: deviceStates[entry.id] ?? .idle
            )
        }
        return AnalyticsRow(
            key: key,
            // Not on this phone and inert either way once there is no entry, so
            // there is nothing for a nil id to act on.
            entryId: entry?.id,
            group: group,
            title: title,
            subtitle: subtitle,
            state: state,
            affordance: affordance
        )
    }

    // A video picked for a match is that match's row, not a standalone one -
    // the same filter `MatchesList`'s own "On this phone" section applies.
    let localVideoRows = localEntries.filter { $0.scoreLogId == nil }.map { entry -> AnalyticsRow in
        return rowFor(
            key: "local-\(entry.id)",
            entryId: entry.id,
            group: .localVideos,
            title: entry.title ?? entry.displayName,
            subtitle: localVideoSubtitle(entry)
        )
    }

    let ownedMatchRows = ownedRows.map { row -> AnalyticsRow in
        switch row {
        case .video(let match):
            return rowFor(
                key: row.id,
                entryId: match.videoId,
                group: .ownedMatches,
                title: matchRowPrimary(match),
                subtitle: matchRowSecondary(match)
            )
        case .score(let content):
            return rowFor(
                key: row.id,
                // The entry that claims this log, which is how the rest of this
                // app pairs a scored match to its video (`scoreRow` in
                // ClipListView, and the shared `scoreLogAttachStatus`). NOT
                // `card.videoId`: that is written only by the cloud pipeline's
                // CREATE_ROW step, so a match scored courtside and filmed on this
                // phone has none, and reading it would call a video sitting on
                // the phone "Not on this phone" forever. Android shipped that
                // version and it was a Critical.
                entryId: localEntries.first { $0.scoreLogId == content.card.scoreLogId }?.id,
                group: .ownedMatches,
                title: content.card.title,
                subtitle: content.card.playersLine
            )
        }
    }

    let sharedRows = sharedMatches.map { match in
        rowFor(
            key: "shared-\(match.videoId)",
            entryId: match.videoId,
            group: .shared,
            title: matchRowPrimary(match),
            subtitle: matchRowSecondary(match)
        )
    }

    // One video, one row. A video-first import the cloud has finished clipping is
    // both a local video and an owned match, and the two rows would carry the
    // same entry id, the same state and two identical "Analyze" buttons - the
    // drawer gets away with that because its two rows look nothing alike, and
    // here they are indistinguishable. First wins, and the order below puts the
    // local row first deliberately: it is the stable one, alive from import until
    // the file leaves the phone, whereas the owned row only appears once a cloud
    // run has clipped, so keeping that one instead would make the row jump
    // sections mid-life. Rows with no entry fall back to their own key, and every
    // NOT_ON_DEVICE row has a nil entryId, so two matches that are merely both
    // absent from this phone never collapse into each other.
    var seen = Set<String>()
    return (localVideoRows + ownedMatchRows + sharedRows).filter {
        seen.insert($0.entryId ?? $0.key).inserted
    }
}

/// Which explanatory line belongs above these rows. See `AnalyticsLegend`.
func analyticsLegend(for rows: [AnalyticsRow]) -> AnalyticsLegend {
    if rows.isEmpty { return .silent }
    // Every row inert for the same reason: nothing on this list has ever touched
    // this phone. Repeating "Not on this phone" down the whole list would say the
    // same thing as many times as there are rows.
    if rows.allSatisfy({ $0.state == .notOnDevice }) { return .nothingOnThisPhone }
    // A dot is a symbol and needs explaining; a button labelled "Analyze" does
    // not, so the dot wins whenever both are on screen.
    if rows.contains(where: { AnalyticsRowStateKt.opensAnalytics(state: $0.state) }) { return .dot }
    return .analyseButton
}

File: androidApp/src/main/java/com/badmintontracker/android/analytics/AnalyticsRows.kt
package com.badmintontracker.android.analytics

import com.badmintontracker.android.cliplist.MatchRow
import com.badmintontracker.android.cliplist.MatchSummary
import com.badmintontracker.android.cliplist.formatDate
import com.badmintontracker.android.cliplist.matchRowPrimary
import com.badmintontracker.android.cliplist.matchRowSecondary
import com.badmintontracker.android.localanalysis.LocalAnalysisState
import com.badmintontracker.android.localvideo.LocalVideoRow
import com.badmintontracker.shared.analytics.AnalyticsRowState
import com.badmintontracker.shared.analytics.analyticsRowState
import com.badmintontracker.shared.localvideo.AnalyzeProgress
import com.badmintontracker.shared.localvideo.AnalyzeStage
import com.badmintontracker.shared.localvideo.DevicePhase
import com.badmintontracker.shared.localvideo.LocalVideoEntry
import com.badmintontracker.shared.localvideo.cloudAnalysisStatus
import com.badmintontracker.shared.localvideo.deviceWorkLabel
import com.badmintontracker.shared.localvideo.isAnalysisRunning
import kotlinx.datetime.Instant

/**
 * Which of the two pipelines a row's control should speak for.
 *
 * Two independent pipelines can be running over the same video at once (see
 * BackgroundWork.kt's own comment on why cloud and on-device failures are
 * tracked apart). The device run wins when both are non-idle, because that is
 * the run this row's own button would start, but the cloud pipeline must not be
 * left silent just because the button starts the other one: a coach watching
 * this list must not see "Analyze" on a video that is, in fact, being analyzed
 * right now, in the cloud.
 */
internal fun affordanceFor(
    entry: LocalVideoEntry,
    live: LocalAnalysisState,
    progress: AnalyzeProgress?,
): AnalyseAffordance = when (live) {
    // Through [deviceWorkLabel] rather than three literals of its own, which is
    // what these used to be: the drawer's row and the chrome indicator already
    // go through it, and a fourth hand-written copy is how this screen kept a
    // spelling the rest of the app had left behind. No fraction, matching the
    // drawer's row - see its comment for why the number is left off a list line.
    is LocalAnalysisState.Preparing -> AnalyseAffordance.InProgress(deviceWorkLabel(DevicePhase.PREPARING, null))
    is LocalAnalysisState.Analysing -> AnalyseAffordance.InProgress(deviceWorkLabel(DevicePhase.ANALYSING, null))
    is LocalAnalysisState.Cutting -> AnalyseAffordance.InProgress(deviceWorkLabel(DevicePhase.CUTTING, null))
    is LocalAnalysisState.Failed -> AnalyseAffordance.Failed(live.message)
    // No device run in flight or failed: fall back to the cloud pipeline's own
    // liveness for this entry, in the shared wording the drawer's row for this
    // same video uses. It used to say "Uploading" and "Processing in the cloud"
    // here, which is the chrome indicator's vocabulary and dropped the
    // percentage - so one tap apart the same run read two ways, and iOS's port
    // of this screen had already gone the other way.
    is LocalAnalysisState.Idle, is LocalAnalysisState.Done -> when {
        // Non-null for both running stages, so the fallback guards a future
        // stage rather than anything reachable today.
        isAnalysisRunning(entry.stage) ->
            AnalyseAffordance.InProgress(cloudAnalysisStatus(entry.stage, progress) ?: "Analyzing…")
        entry.stage == AnalyzeStage.FAILED -> AnalyseAffordance.Failed(entry.failureMessage ?: "Unknown error")
        // A track only lands on disk when the run asked for pose (see
        // LocalAnalysisRunner.start): a Done device run that did not is not a
        // liveness problem, just a video still waiting for its first (pose)
        // analysis - the same as one that was never touched.
        else -> AnalyseAffordance.Ready
    }
}

/**
 * Which explanatory line sits above the list, if any.
 *
 * The screen used to switch two ways: all-inert, or the dot legend. That put
 * "Analyzed on this phone - tap to view" and a green dot above a list where the
 * only row said "Analyze" and no dot was drawn anywhere, because ANALYSABLE
 * fails the all-inert test without contributing a dot. A legend explaining a
 * symbol that is not on screen is worse than no legend. Found on iOS, where
 * every local video is ANALYSABLE and it was the ONLY thing a coach would see.
 *
 * A dot is a symbol and needs explaining; a button already says what it does,
 * so the dot wins whenever both are present.
 */
internal enum class AnalyticsLegend {
    SILENT, NOTHING_ON_THIS_PHONE, DOT, ANALYSE_BUTTON;

    /** Whether any row draws the availability dot the [DOT] line explains. */
    val showsDot: Boolean get() = this == DOT

    /**
     * Whether an inert row spells out "Not on this phone" for itself.
     *
     * False only when the legend already said it once for the whole list. Kept
     * here rather than in the view so the legend and the rows cannot disagree
     * about which of them is doing the explaining.
     */
    val showsNotOnDeviceSubtitle: Boolean get() = this != NOTHING_ON_THIS_PHONE
}

internal fun analyticsLegend(rows: List<AnalyticsRow>): AnalyticsLegend = when {
    rows.isEmpty() -> AnalyticsLegend.SILENT
    // Repeating "Not on this phone" down the whole list would say the same thing
    // as many times as there are rows.
    rows.all { it.state == AnalyticsRowState.NOT_ON_DEVICE } -> AnalyticsLegend.NOTHING_ON_THIS_PHONE
    rows.any { it.state == AnalyticsRowState.READY } -> AnalyticsLegend.DOT
    else -> AnalyticsLegend.ANALYSE_BUTTON
}

/**
 * Builds the Analytics list's rows, grouped like the drawer: local videos, then
 * owned matches, then shared. [analyticsRowState] alone decides READY /
 * ANALYSABLE / NOT_ON_DEVICE - this only gathers its two booleans per match
 * and, for an ANALYSABLE one, reads both pipelines' own liveness via
 * [affordanceFor] so the row can show progress or a failure instead of a live
 * "Analyze" button. `hasStoredTrack` alone cannot tell a run in flight, or one
 * that just failed, from one never attempted.
 *
 * [storedTrackIds] is a caller-computed set rather than a per-row disk probe:
 * this function runs on every recomposition, including the several-times-a-
 * second ones an in-flight analysis's progress causes.
 */
internal fun buildAnalyticsRows(
    standaloneLocalRows: List<LocalVideoRow>,
    ownedRows: List<MatchRow>,
    sharedMatches: List<MatchSummary>,
    localEntries: List<LocalVideoEntry>,
    liveAnalysisStates: Map<String, LocalAnalysisState>,
    progressByEntryId: Map<String, AnalyzeProgress>,
    storedTrackIds: Set<String>,
): List<AnalyticsRow> {
    fun rowFor(key: String, entryId: String?, group: AnalyticsGroup, title: String, subtitle: String): AnalyticsRow {
        val entry = entryId?.let { id -> localEntries.firstOrNull { it.id == id } }
        val hasStoredTrack = entryId != null && entryId in storedTrackIds
        val state = analyticsRowState(hasLocalEntry = entry != null, hasStoredTrack = hasStoredTrack)
        // Only an ANALYSABLE row reads liveness, and that is a decision rather
        // than an oversight. A READY row whose second run is in flight keeps its
        // dot and stays tappable, because the track the first run produced still
        // opens - swapping the dot for a spinner would take away a working
        // affordance to report a run whose result the row does not need yet.
        // LocalVideoRowItem spins whenever isAnalysisRunning; it has no stored
        // result to offer instead, so it has nothing to lose by doing that.
        //
        // entry is never null here: ANALYSABLE requires hasLocalEntry, which is
        // exactly `entry != null` above. The null check stays as a guard, not a
        // second source of truth, so this cannot throw if that ever changes.
        val affordance = if (state == AnalyticsRowState.ANALYSABLE && entry != null) {
            affordanceFor(
                entry,
                liveAnalysisStates[entry.id] ?: LocalAnalysisState.Idle,
                progressByEntryId[entry.id],
            )
        } else {
            AnalyseAffordance.Ready
        }
        return AnalyticsRow(
            key = key,
            // Not on this phone and inert either way once there is no entry,
            // so there is nothing for a null id to navigate to.
            entryId = entry?.id,
            group = group,
            title = title,
            subtitle = subtitle,
            state = state,
            affordance = affordance,
        )
    }

    val localVideoRows = standaloneLocalRows.map { row ->
        rowFor(
            key = "local-${row.entry.id}",
            entryId = row.entry.id,
            group = AnalyticsGroup.LOCAL_VIDEOS,
            title = row.primaryText,
            // Sentence case, as the mock has it ("1:05 · Sep 2"). This screen
            // used to uppercase its match rows and not its local video rows, so
            // one list drew "0:02 · Sep 4, 2026" directly over
            // "2 RALLIES · SEP 1, 2026"; the answer turned out to be neither.
            subtitle = "${row.durationText} · " +
                formatDate(Instant.fromEpochMilliseconds(row.entry.addedAtEpochMs)),
        )
    }

    val ownedMatchRows = ownedRows.map { row ->
        when (row) {
            is MatchRow.Video -> rowFor(
                key = row.key,
                entryId = row.match.videoId,
                group = AnalyticsGroup.OWNED_MATCHES,
                title = matchRowPrimary(row.match),
                subtitle = matchRowSecondary(row.match),
            )
            is MatchRow.Score -> rowFor(
                key = row.key,
                // The entry that claims this log, which is how the rest of the
                // app pairs a scored match to its video (scoreLogAttachStatus
                // in AttachStatus.kt, and both of AuthGate's attached-video
                // lookups). Not `card.videoId`: that is written only by the
                // CLOUD pipeline's CREATE_ROW hook, so a match scored courtside
                // and filmed on this phone has none, and reading it would call
                // a video sitting on the phone "Not on this phone" forever.
                entryId = localEntries.firstOrNull { it.scoreLogId == row.card.scoreLogId }?.id,
                group = AnalyticsGroup.OWNED_MATCHES,
                title = row.card.title,
                subtitle = row.card.playersLine,
            )
        }
    }

    val sharedRows = sharedMatches.map { match ->
        rowFor(
            key = "shared-${match.videoId}",
            entryId = match.videoId,
            group = AnalyticsGroup.SHARED,
            title = matchRowPrimary(match),
            subtitle = matchRowSecondary(match),
        )
    }

    // One video, one row. A video-first import the cloud has finished clipping
    // is both a local video and an owned match, and the two rows carry the same
    // entry id, the same state and two identical "Analyze" buttons - the drawer
    // gets away with that because its two rows look nothing alike, and here they
    // are indistinguishable. The local row wins because it is the stable one: it
    // exists from import until the file leaves the phone, whereas the owned row
    // appears only once a cloud run finishes, so keeping that one instead would
    // make the row jump sections mid-life. Rows with no entry fall back to their
    // own key, and every NOT_ON_DEVICE row has a null entryId, so two matches
    // that are merely both absent from this phone never collapse into each other.
    return (localVideoRows + ownedMatchRows + sharedRows).distinctBy { it.entryId ?: it.key }
}
